direct.go

```diff
@@ -411,10 +410,35 @@ func (d *direct) doDial(m *Masquerade) (conn net.Conn, retriable bool, err error
 }
 
 func (d *direct) dialServerWith(m *Masquerade) (net.Conn, error) {
+	op := ops.Begin("dial_server_with")
+	defer op.End()
+
+	op.Set("masquerade_domain", m.Domain)
+	op.Set("masquerade_ip", m.IpAddress)
+
 	tlsConfig := d.frontingTLSConfig(m)
 	dialTimeout := 10 * time.Second
-	sendServerNameExtension := false
 	addr := m.IpAddress
+	var sendServerNameExtension bool
+
+	// looking for provider and using SNI if enabled
+	provider := d.findProviderFromMasquerade(m)
+	if provider != nil && provider.SNIConfig != nil && provider.SNIConfig.UseArbitrarySNIs {
+		sendServerNameExtension = true
+
+		// selecting a random SNI
+		randomSNIIndex := rand.IntN(len(provider.SNIConfig.ArbitrarySNIs))
+		sniDomain := provider.SNIConfig.ArbitrarySNIs[randomSNIIndex]
```

Comment:
I think we actually want this to be deterministic - and deterministic across clients too.  The ideal scenario IMO is if any two clients in region R both send requests to IP X, they both use the same SNI.  So perhaps we hash the masquerade IP and use the result modulo the set of SNIs to choose an SNI.  Does that make sense?

A couple of minor points with my scheme: (1) If two clients have different lists, they will not end up with the same SNI.  That is okay though; I think clients should only end up with different lists if they're in different regions.  In that case, we don't need them to use the same SNI.  (2) If the SNI list changes, then SNI-to-IP mappings will change.  This is okay too; I don't expect us to change the SNI often.

Sorry, I know this is a new requirement added after the fact.  I realize that can be frustrating.  This didn't occur to me earlier =/

Reply:
I've added a `crc32.ChecksumIEEE` func call for hashing and made some changes for distributing SNIs per masquerade.